server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java

```diff
@@ -523,30 +526,44 @@ protected String doReadString(final int charCount) throws IOException {
         return new String(charBuffer, 0, charCount);
     }
 
-    protected static int calculateByteLengthOfChars(final byte[] bytes, final int chars, final int start, final int limit)
-        throws IOException {
-        if (limit - start < chars) {
-            return -1; // not enough bytes to read chars
+    protected String tryReadStringFromBytes(final byte[] bytes, final int start, final int limit, final int chars) throws IOException {
+        final int end = start + chars;
+        if (limit < end) {
+            return null; // not enough bytes to read chars
         }
-        // calculate number of bytes matching the requested chars
-        int pos = start;
-        int remaining = chars;
-        while (pos < limit && remaining-- > 0) {
+        for (int pos = start; pos < end; pos++) {
+            if ((bytes[pos] & 0xf0) > 0x70) {
```

Comment:
I'm curious, is this better than `(bytes[pos] & 0x80) != 0` in some way? Deserves a comment I think.

Reply:
that just originated from `(bytes[pos] & 0xff) >> 4 > 7` without much thinking, just checking the msb is obviously simpler. I'll change, thanks :)